plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/AbstractTestOracleTypeMapping.java

```diff
@@ -733,6 +734,50 @@ public void testTimestamp(ZoneId sessionZone)
                 .execute(getQueryRunner(), session, trinoCreateAndInsert(session, "test_timestamp"));
     }
 
+    @Test(dataProvider = "sessionZonesDataProvider")
+    public void testTimestampNanos(ZoneId sessionZone)
+    {
+        Session session = Session.builder(getSession())
+                .setTimeZoneKey(getTimeZoneKey(sessionZone.getId()))
+                .build();
+
+        SqlDataTypeTest.create()
+                // min value in Oracle
+                .addRoundTrip("TIMESTAMP(9)", "TIMESTAMP '-4712-01-01 00:00:00.000000000'", TIMESTAMP_NANOS, "TIMESTAMP '-4712-01-01 00:00:00.000000000'")
+                .addRoundTrip("TIMESTAMP(9)", "TIMESTAMP '-0001-01-01 00:00:00.000000000'", TIMESTAMP_NANOS, "TIMESTAMP '-0001-01-01 00:00:00.000000000'")
+                // day before and after julian->gregorian calendar switch
+                .addRoundTrip("TIMESTAMP(9)", "TIMESTAMP '1582-10-04 00:00:00.000000000'", TIMESTAMP_NANOS, "TIMESTAMP '1582-10-04 00:00:00.000000000'")
+                .addRoundTrip("TIMESTAMP(9)", "TIMESTAMP '1582-10-15 00:00:00.000000000'", TIMESTAMP_NANOS, "TIMESTAMP '1582-10-15 00:00:00.000000000'")
+                // before epoch
+                .addRoundTrip("TIMESTAMP(9)", "TIMESTAMP '1958-01-01 13:18:03.123123123'", TIMESTAMP_NANOS, "TIMESTAMP '1958-01-01 13:18:03.123123123'")
+                // after epoch
+                .addRoundTrip("TIMESTAMP(9)", "TIMESTAMP '2019-03-18 10:01:17.987987987'", TIMESTAMP_NANOS, "TIMESTAMP '2019-03-18 10:01:17.987987987'")
+                // epoch, epoch also is a gap in JVM zone
+                .addRoundTrip("TIMESTAMP(9)", "TIMESTAMP '1970-01-01 00:00:00.000000000'", TIMESTAMP_NANOS, "TIMESTAMP '1970-01-01 00:00:00.000000000'")
+                .addRoundTrip("TIMESTAMP(9)", timestampDataType(9).toLiteral(timeDoubledInJvmZone), TIMESTAMP_NANOS, timestampDataType(9).toLiteral(timeDoubledInJvmZone))
+                .addRoundTrip("TIMESTAMP(9)", timestampDataType(9).toLiteral(timeDoubledInVilnius), TIMESTAMP_NANOS, timestampDataType(9).toLiteral(timeDoubledInVilnius))
+                .addRoundTrip("TIMESTAMP(9)", timestampDataType(9).toLiteral(timeGapInJvmZone1), TIMESTAMP_NANOS, timestampDataType(9).toLiteral(timeGapInJvmZone1))
+                .addRoundTrip("TIMESTAMP(9)", timestampDataType(9).toLiteral(timeGapInJvmZone2), TIMESTAMP_NANOS, timestampDataType(9).toLiteral(timeGapInJvmZone2))
+                .addRoundTrip("TIMESTAMP(9)", timestampDataType(9).toLiteral(timeGapInVilnius), TIMESTAMP_NANOS, timestampDataType(9).toLiteral(timeGapInVilnius))
+                .addRoundTrip("TIMESTAMP(9)", timestampDataType(9).toLiteral(timeGapInKathmandu), TIMESTAMP_NANOS, timestampDataType(9).toLiteral(timeGapInKathmandu))
+                // max value in Oracle
+                .addRoundTrip("TIMESTAMP(9)", "TIMESTAMP '9999-12-31 00:00:00.000000000'", TIMESTAMP_NANOS, "TIMESTAMP '9999-12-31 00:00:00.000000000'")
```

Comment:
Oracle doesn't support `TIMESTAMP '9999-12-31 23:59:59.999999999'`? Why all hour, min and sec are 0?

Reply:
added